### ISSUE TITLE
Translate Score DSL expressions to natural language

### ISSUE THREAD

**Issue by donmendelson** [FIXTradingCommunity/fix-orchestra#13]
Although the Score DSL grammar is very familiar to programmers, it should be translated to natural language for non-technical users in visualizations of Orchestra. This applies to docgen and testgen as well as other tools developed in the future that will have a UI.

Most translations can be simple substitution of tokens, e.g. "is less than" in place of the token "<".  Ideally, the substitution should be table-driven so expressions can be rendered in multiple languages.

**Comment by donmendelson** [comment]
It's not hard to substitute words for symbols in a formulate like "times" for "*". I have found that the awkward part is how to describe nested message elements in words, e.g. the PartyRole field of the fourth entry of the Parties repeating group. I'm looking for suggestions.

**Comment by kleihan** [comment]
I use the term "instance" and not "entry" and a shorter expression, i.e. "PartyRole of the fourth instance of Parties". The term "instance of" implies that the name of a repeating group will follow. In writing, one can additionally set the name into angle brackets and add a tag number to imply a field, i.e. "PartyRole(452) of the fourth instance of <Parties>".

**Comment by donmendelson** [comment]
Natural language processing is advancing very rapidly. However, new technologies require very large training databases, which seems highly unlikely for our specialized grammar.

**Comment by jimnup** [comment]
We should go back and look at natural language rules languages before we
reinvent. There have been a few efforts, some of them have been around for
quite some time. None have gained wide traction.

As far as NLP, my overarching concern is people seem to be turning to
robotic process automation (RPA) and machine learning (ML) including NLP in
place of building appropriate systems with structured information. When is
the use of RPA and NLP putting bandages or in the US vernacular "spit and
bailing wire" and when are we expanding into uses of AI that extend beyond
traditional systems. Tough calls these.

Anyway, I think a natural language simplification that could be dictionary
driven as opposed to the full force of NLP would make sense and in keeping
with the power we are seeing with tablature.

**Comment by donmendelson** [comment]
@jimnup I agree that a dictionary and rule based approach is more appropriate, and Orchestra provides a dictionary of message, field, and code names as objects to operate on. What is needed is natural language terms for relationships between the objects, e.g. knowledge that a code name is in the domain a field that is bound to a certain codeset. Useful phrases will never be as quite simple as a word-wise dictionary lookup, but the scope can be much more contained than free-form translation provided by neural networks.
